File: src/uyap_mcp/worker.py
"""Tüm portal işlemlerini tek Playwright iş parçacığında sıraya koyar."""

from __future__ import annotations

import atexit
import logging
from concurrent.futures import ThreadPoolExecutor
from typing import Any, Callable

from .config import Settings
from .session import UyapSession

LOGGER = logging.getLogger("uyap_mcp")


class PortalWorker:
    """MCP çağrılarının UYAP tarafında birbiriyle yarışmasını engeller."""

    def __init__(self, settings: Settings | None = None):
        self.settings = settings or Settings.from_env()
        self._executor = ThreadPoolExecutor(max_workers=1, thread_name_prefix="uyap-portal")
        self._session: UyapSession | None = None
        self._closed = False
        atexit.register(self.close)

    def _get_session(self) -> UyapSession:
        if self._session is None:
            self._session = UyapSession(self.settings, LOGGER.info)
        return self._session

    def call(self, method: str, *args: Any, **kwargs: Any) -> Any:
        def invoke() -> Any:
            target: Callable[..., Any] = getattr(self._get_session(), method)
            return target(*args, **kwargs)

        return self._executor.submit(invoke).result()

    def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        if self._session is not None:
            try:
                self._executor.submit(self._session.close).result(timeout=5)
            except RuntimeError:
                # Python, atexit işleyicilerinden önce executor havuzunu kapatmış olabilir.
                pass
            except Exception:
                LOGGER.exception("Portal oturumu kapatılırken hata oluştu.")
            self._session = None
        self._executor.shutdown(wait=False, cancel_futures=True)

File: src/services/worker.ts
/** Tüm portal işlemlerini tek sıralı kuyrukta çalıştırır. */

import type { Settings } from "../config.js";
import {
  UyapSession,
  type DocumentsResult,
  type DownloadResult,
  type PortalCase,
  type SessionStatus,
} from "./session.js";

function log(message: string): void {
  console.error(`INFO uyap_mcp: ${message}`);
}

/** MCP çağrılarının UYAP tarafında birbiriyle yarışmasını engeller. */
export class PortalWorker {
  private readonly settings: Settings;
  private session: UyapSession | null = null;
  private queue: Promise<unknown> = Promise.resolve();
  private closed = false;

  constructor(settings: Settings) {
    this.settings = settings;
  }

  private getSession(): UyapSession {
    if (!this.session) {
      this.session = new UyapSession(this.settings, log);
    }
    return this.session;
  }

  private enqueue<T>(task: (session: UyapSession) => Promise<T>): Promise<T> {
    const result = this.queue.then(() => task(this.getSession()));
    this.queue = result.then(
      () => undefined,
      () => undefined
    );
    return result;
  }

  status(): Promise<SessionStatus> {
    return this.enqueue((session) => session.status());
  }

  prepareLogin(waitSeconds: number): Promise<SessionStatus> {
    return this.enqueue((session) => session.prepareLogin(waitSeconds));
  }

  listCases(status: 0 | 1): Promise<PortalCase[]> {
    return this.enqueue((session) => session.listCases(status));
  }

  listDocuments(caseNo: string, unit: string | null | undefined, status: 0 | 1): Promise<DocumentsResult> {
    return this.enqueue((session) => session.listDocuments(caseNo, unit, status));
  }

  downloadCase(caseNo: string, unit: string | null | undefined, status: 0 | 1): Promise<DownloadResult> {
    return this.enqueue((session) => session.downloadCase(caseNo, unit, status));
  }

  async close(): Promise<void> {
    if (this.closed) return;
    this.closed = true;
    if (this.session) {
      const session = this.session;
      this.session = null;
      await session.close().catch(() => undefined);
    }
  }
}
